pontoon/pretranslation/tasks.py

```diff
@@ -0,0 +1,115 @@
+import logging
+
+from django.utils import timezone
+from django.db.models import Q, CharField, Value as V
+from django.db.models.functions import Concat
+
+import pontoon.base as base
+from pontoon.pretranslation.pretranslate import get_translations, update_changed_instances
+
+
+log = logging.getLogger(__name__)
+
+
+def pretranslate(project, locales=None, entities=None):
+    """
+    Identifies strings without any translations and any suggestions.
+    Engages TheAlgorithm (bug 1552796) to gather pretranslations.
+    Stores pretranslations as suggestions (approved=False) to DB.
+
+    :arg Project project: the project to be pretranslated
+    :arg Queryset locales: the locales for the project to be pretranslated
+    :arg Queryset entites: the entities for the project to be pretranslated
+
+    :returns: None
+    """
+
+    log.info('Fetching pretranslations for project {} started'.format(project.name))
+
+    if not locales:
+        locales = project.locales.filter(
+            project_locale__readonly=False,
+        ).prefetch_project_locale(project)
+
+    if not entities:
+        entities = base.models.Entity.objects.filter(
+            resource__project=project,
+            obsolete=False,
+        ).prefetch_related('resource')
+
+    # Fetch all distinct locale-entity pairs for which translation exists
+    translated_entities = base.models.Translation.objects.filter(
+        locale__in=locales,
+        entity__in=entities,
+    ).annotate(
+        locale_entity=Concat('locale_id', V('-'), 'entity_id', output_field=CharField())
+    ).values_list('locale_entity', flat=True).distinct()
+
+    translated_entities = list(translated_entities)
+
+    now = timezone.now()
+
+    translations = []
+
+    # To keep track of changed Locales and TranslatedResources
+    # Also, their latest_translation and stats count
+    locale_dict = {}
+    tr_dict = {}
+
+    tr_filter = []
+    index = -1
+
+    for locale in locales:
+        log.info('Fetching pretranslations for locale {} started'.format(locale.code))
+        for entity in entities:
+            locale_entity = '{}-{}'.format(locale.id, entity.id)
+            if locale_entity not in translated_entities:
+                strings = get_translations(entity, locale)
+                if strings:
+                    for string, plural_form, user in strings:
+                        t = base.models.Translation(
+                            entity=entity,
+                            locale=locale,
+                            string=string,
+                            user=user,
+                            date=now,
```

Comment:
That means all translations across all locales and entities will have the same timestamp. Which will also be reflected in the latest activity column. Is that intentional?

`Translate.date` sets the default value to the actual `timezone.now`, why don't you just keep that (by not setting the date here at all)?

Reply:
We will leave it default and I will change the latest translation to last one as well